### PR TITLE
fix: handle session and confirm flow discrepancy in surcharge details

### PR DESCRIPTION
**Pull request by hrithikesh026**
## Type of Change

- [x] Bugfix
- [ ] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

In case of wallet payment method types like google pay and apple pay, before confirm call, we will have to do a session call. 
It may  be possible that in this session call, we will be sending the original amount to the connector and later in confirm call, the merchant can send some surcharge amount. This will cause some problems since, amount being sent in session call and confirm call will differ.

Above issue is solved in this PR by following steps:
1. Merchant can send some surcharge details during payments create. This will be stored in `payment_attempt.surcharge_amount` and `payment_attempt.tax_amount`.
2. in session flow, we will send the original_amount+surcharge+tax_amount to the connector .
3. During confirm call, if there is some surcharge_details in request and payment method is the one that requires session token, we will verify that amount sent in session call is same as total amount we will be sending in confirm. If the amounts do not match, We throw an error saying "session call has already being done with a different amount for this payment method type".

Other Changes:
Enable updating `surcharge_amount` and `tax_amount` in payment_update api.
Fixed `amount_captured` being original amount during `auto_capture` payment even when surcharge is available. 

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?

Manual

Pass surcharge in Payment create. Will get updated in payment_attempt
Screenshot 2023-10-26 at 12 32 54 PM

Screenshot 2023-10-26 at 12 34 46 PM

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/core/payments.rs [hunk 1]
```diff
@@ -258,6 +266,69 @@ where
     ))
 }
 
+pub async fn carry_out_session_surcharge_operations(
+    state: &AppState,
+    connectors: &Vec<api::SessionConnectorData>,
```
Comment (Narayanbhat166) [inline crates/router/src/core/payments.rs:271]: name it as session connectors data

**Review by Narayanbhat166 — APPROVED** [review]

File: crates/router/src/core/utils.rs [hunk 2]
```diff
@@ -1073,3 +1079,57 @@ pub fn get_flow_name<F>() -> RouterResult<String> {
...
+    let redis_store_futures_iter = surcharge_metadata
+        .get_individual_surcharge_key_value_pairs()
+        .into_iter()
+        .map(|(redis_key, surcharge_details)| {
+            redis_conn.serialize_and_set_key_with_expiry(
+                redis_key,
+                surcharge_details,
+                merchant_account
+                    .intent_fulfillment_time
+                    .unwrap_or(DEFAULT_FULFILLMENT_TIME),
```
Comment (Narayanbhat166) [inline crates/router/src/core/utils.rs:1103]: use `set_hash_fields` redis function here. The key would be `surcharge_metadata_<attempt_id>` and the individual fields can be hashmap keys.

File: crates/router/src/types/api.rs [hunk 3]
```diff
@@ -213,6 +214,14 @@ pub struct SessionConnectorData {
     pub business_sub_label: Option<String>,
 }
 
+/// Session Surcharge type
+pub enum SessionSurchargeDetails {
+    /// Surcharge is calculated by us
```
Comment (Narayanbhat166) [inline crates/router/src/types/api.rs:219]:
```suggestion
    /// Surcharge is calculated by hyperswitch
```

File: crates/router/src/core/payments/operations/payment_confirm.rs [hunk 4]
```diff
@@ -699,3 +721,79 @@ impl<F: Send + Clone, Ctx: PaymentMethodRetrieve> ValidateRequest<F, api::Paymen
...
+                            Err(err) if err.current_context() == &RedisError::NotFound => {
+                                if request_surcharge_details.surcharge_amount == 0
+                                    && request_surcharge_details.tax_amount.unwrap_or(0) == 0
+                                {
+                                    return Ok(());
+                                } else {
+                                    return invalid_surcharge_details_error;
+                                }
+                            }
+                            Err(_) => {
```
Comment (jarnura) [inline crates/router/src/core/payments/operations/payment_confirm.rs:781]: don't discard the error, lift the into InternalServerError.

File: crates/router/src/core/payments/operations/payment_confirm.rs [hunk 5]
```diff
@@ -699,3 +721,79 @@ impl<F: Send + Clone, Ctx: PaymentMethodRetrieve> ValidateRequest<F, api::Paymen
...
+                            state,
+                            &payment_method_type.into(),
+                            &payment_method_type,
+                            None,
+                            &payment_attempt.attempt_id,
+                        )
+                        .await
+                        {
+                            Ok(surcharge_details) => {
+                                if surcharge_details
```
Comment (jarnura) [inline crates/router/src/core/payments/operations/payment_confirm.rs:764]: when function can be used here. Below if else block it is applicable.

**Review by jarnura — CHANGES_REQUESTED** [review]

**Review by jarnura — APPROVED** [review]

File: crates/router/src/types/api.rs [hunk 6]
```diff
@@ -214,6 +215,30 @@ pub struct SessionConnectorData {
...
+impl SessionSurchargeDetails {
+    pub fn get_surcharge_details(
+        &self,
+        payment_method: &enums::PaymentMethod,
+        payment_method_type: &enums::PaymentMethodType,
+        card_network: Option<&enums::CardNetwork>,
+    ) -> Option<SurchargeDetailsResponse> {
+        match self {
+            Self::Calculated(surcharge_metadata) => surcharge_metadata
+                .get_surcharge_details(payment_method, payment_method_type, card_network)
```
Comment (Narayanbhat166) [inline crates/router/src/types/api.rs:235]: Please avoid having two functions with the same name

**Review by Narayanbhat166 — APPROVED** [review]

**Review by jarnura — APPROVED** [review]

**Review by jarnura — DISMISSED** [review]

**Review by Narayanbhat166 — DISMISSED** [review]

**Review by jarnura — DISMISSED** [review]

**Review by Narayanbhat166 — DISMISSED** [review]

**Review by jarnura — APPROVED** [review]

File: crates/router/src/core/payments/operations/payment_confirm.rs [hunk 7]
```diff
@@ -672,3 +665,92 @@ impl<F: Send + Clone, Ctx: PaymentMethodRetrieve> ValidateRequest<F, api::Paymen
...
+            request.payment_method_data.as_ref(),
+        ) {
+            (Some(request_surcharge_details), Some(payment_method_data)) => {
+                if let Some(payment_method_type) =
+                    payment_method_data.get_payment_method_type_if_session_token_type()
+                {
+                    let invalid_surcharge_details_error = Err(errors::ApiErrorResponse::InvalidRequestData {
+                        message: "surcharge_details sent in session token flow doesn't match with the one sent in confirm request".into(),
+                    }.into());
+                    if let Some(attempt_surcharge_amount) = payment_attempt.surcharge_amount {
```
Comment (Narayanbhat166) [inline crates/router/src/core/payments/operations/payment_confirm.rs:686]: Can you make these two if lets into a single match statement?

File: crates/router/src/core/utils.rs [hunk 8]
```diff
@@ -1073,3 +1081,65 @@ pub fn get_flow_name<F>() -> RouterResult<String> {
...
+            .get_individual_surcharge_key_value_pairs()
+            .into_iter()
+        {
+            value_list.push((
+                key,
+                Encode::<SurchargeDetailsResponse>::encode_to_string_of_json(&value)
+                    .change_context(errors::ApiErrorResponse::InternalServerError)
+                    .attach_printable("Failed to encode to string of json")?,
+            ));
+        }
```
[the remark this answers is not in the document]
Comment (hrithikesh026) [reply crates/router/src/core/utils.rs:1111]: This function  `surcharge_metadata.get_individual_surcharge_key_value_pairs(`) returns a `Vec<(String, SurchargeDetails)>`.

I am just converting it into Vec<(String, String)> inside for loop.

**Comment by Narayanbhat166** [comment]
https://github.com/juspay/hyperswitch/actions/runs/6854446608/job/18637405889?pr=2696#step:19:436
Can you please address this?

**Review by jarnura — APPROVED** [review]

**Review by Narayanbhat166 — APPROVED** [review]

**Pull request merged**